Content.Server/Speech/EntitySystems/IlleismAccentSystem.cs

```diff
@@ -0,0 +1,122 @@
+using Content.Server.Speech.Components;
+
+namespace Content.Server.Speech.EntitySystems;
+
+public sealed class IlleismAccentSystem : EntitySystem
+{
+    // I am going to Sec -> NAME is going to Sec
+    private static readonly Regex RegexIAmUpper = new(@"\bI\s*AM\b");
+    private static readonly Regex RegexIAmLower = new(@"\bi\s*am\b", RegexOptions.IgnoreCase);
+
+    // I have it -> NAME has it
+    private static readonly Regex RegexIHaveUpper = new(@"\bI\s*HAVE\b");
+    private static readonly Regex RegexIHaveLower = new(@"\bi\s*have\b", RegexOptions.IgnoreCase);
+
+    // I do! -> NAME does!
+    private static readonly Regex RegexIDoUpper = new(@"\bI\s*DO\b");
+    private static readonly Regex RegexIDoLower = new(@"\bi\s*do\b", RegexOptions.IgnoreCase);
+
+    // I don't! -> NAME doesn't!
+    private static readonly Regex RegexIDontUpper = new(@"\bI\s+DON'?T\b");
+    private static readonly Regex RegexIDontLower = new(@"\bi\s+don'?t\b", RegexOptions.IgnoreCase);
+
+    // I/Myself -> NAME
+    private static readonly Regex RegexMyselfUpper = new(@"\bMYSELF\b");
+    private static readonly Regex RegexI = new(@"\bI\b|\bmyself\b", RegexOptions.IgnoreCase);
+
+    // Me -> NAME
+    private static readonly Regex RegexMeUpper = new(@"\bME\b");
+    private static readonly Regex RegexMeLower = new(@"\bme\b", RegexOptions.IgnoreCase);
+
+    // My crowbar -> NAME's crowbar
+    // That's mine! -> That's NAME's
+    private static readonly Regex RegexMyUpper = new(@"\bMY\b|\bMINE\b");
+    private static readonly Regex RegexMyLower = new(@"\bmy\b|\bmine\b", RegexOptions.IgnoreCase);
+
+    // I'll do it -> NAME'll do it
+    private static readonly Regex RegexIllUpper = new(@"\bI'?LL\b");
+    private static readonly Regex RegexIllLower = new(@"\bi'?ll\b", RegexOptions.IgnoreCase);
+
+
+    [Dependency] private readonly ReplacementAccentSystem _replacement = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<IlleismAccentComponent, AccentGetEvent>(OnAccent);
+    }
+
+    private bool MostlyUppercase(string message)
+    {
+        int totalLetters = 0;
+        int uppercaseLetters = 0;
+
+        // Iterate through each character in the string
+        foreach (char c in message)
+        {
+            if (char.IsLetter(c)) // Check if the character is a letter
+            {
+                totalLetters++;
+                if (char.IsUpper(c)) // Check if the letter is uppercase
+                {
+                    uppercaseLetters++;
+                }
+            }
+        }
+        if (totalLetters == 0)
+        {
+            return false;
+        }
+        return uppercaseLetters > totalLetters / 2;
+    }
+
+    private void OnAccent(EntityUid uid, IlleismAccentComponent component, AccentGetEvent args)
+    {
+        var entityManager = IoCManager.Resolve<IEntityManager>();
+        var message = args.Message;
+        var name = Name(uid).Split(' ')[0];
```

Comment:
It'd be nice if we would not create sub strings for each space character in name (use Span magics), but that would be kinda questionable micro-optimization and a lot of additional code for quite rare scenario. I'd better listen to ideas of maintainers on this.

Reply:
I agree. There's surely a way to only split at the first space rather than every space or to grab up until the first space without splitting at all. The only concern with that, as you brought up, is considering if that tiny optimization makes a big enough difference to warrant the increase in code. I'll brainstorm some options and get maintainer consideration as well.